core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorUtil.java

```diff
@@ -1326,6 +1327,48 @@ public static void adjustTypeForArrayConstructor(
     }
   }
 
+  /**
+   * When the array_append and array_prepend element does not equal the array component type
+   * make explicit casting.
+   *
+   * @param componentType derived array component type
+   * @param opBinding description of call
+   */
```

Comment:
There is no need to implement this. For the form of `array_append/array_prepend/array_insert`, the array of the first arg has been cast. The only thing you need to do is to cast the elementType into the component type.